doc/dev/architecture: life of a repository (#9627)
sourcegraph/sourcegraph: doc/dev/architecture/life-of-a-repository.md

# Life of a repository

This document describes how our backend systems clone and update repositories from a code host.

> ## High Level

Reviewer: nit: I think we don't title each word in headings everywhere else (same for other headings in this document).

Author: ```suggestion
## High level
```

1. An admin configures a [code host configuration](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24%40v3.14.0+file:%5Eschema/%28aws%7Cbit%7Cgit%7Cother%29.*schema%5C.json%24&patternType=literal).

> 2. `repo-updater` periodically [syncs](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/syncer.go#L101) all repository metadata from configured codehosts

Reviewer: ```suggestion
2. `repo-updater` periodically [syncs](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/syncer.go#L101) all repository metadata from configured code hosts.
```

  1. We [poll](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/syncer.go#L354:18) the code host's API based on the configuration.

>   2. We [add/update/remove](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/syncer.go#L142-147) entries in our [`repos` table](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/frontend/db/schema.md#table-public-repo).

Reviewer: ```suggestion
  2. We [add/update/remove](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/syncer.go#L142-147) entries in our [`repo` table](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/frontend/db/schema.md#table-public-repo).
```

> 3. All repositories in our `repos` table are in a [scheduler](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/scheduler.go#L82-95) on `repo-updater` which ensures they are cloned and updated on [`gitserver`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/gitserver/server/server.go#L385:18).

Reviewer: ```suggestion
3. All repositories in our `repo` table are in a [scheduler](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/scheduler.go#L82-95) on `repo-updater` which ensures they are cloned and updated on [`gitserver`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/gitserver/server/server.go#L385:18).
```

> Our guiding principle is to ensure all repositories configured by an administrator are cloned and up to date. However, we need to avoid overloading a code host with API and Git requests.

Reviewer: I think we prepend the term "site" almost everywhere when we refer to "admin" or "administrator".
```suggestion
Our guiding principle is to ensure all repositories configured by a site administrator are cloned and up to date. However, we need to avoid overloading a code host with API and Git requests.
```

Note: Sourcegraph.com is different since it isn't feasible to maintain a clone of all open source repositories. It works via on-demand requests from users.

> Note: There is one other way repositories are fetched. A new commit may not be on Sourcegraph yet, but a user is browsing it via our browser extension. `gitserver` supports a ["EnsureRevision"](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/gitserver/server/server.go#L645) use-case which will do a "git fetch" for the missing revision.

Reviewer: `>NOTE: ...`

Creates an (blue) information box on docs.sourcegraph.com, if that's what you want.

## Services

> `repo-updater` is responsible for communicating with Code Host APIs and co-ordinating the state we synchronise from them. It is a singleton service. It is responsible for maintaining the `repos` table which other services read. It is also responsible for scheduling clones/fetches on `gitserver`. It is also responsible for anything which communicates with a Code Host API. So our campaigns syncer also lives in `repo-updater`.

Reviewer: ```suggestion
`repo-updater` is responsible for communicating with Code Host APIs and co-ordinating the state we synchronise from them. It is a singleton service. It is responsible for maintaining the `repo` table which other services read. It is also responsible for scheduling clones/fetches on `gitserver`. It is also responsible for anything which communicates with a Code Host API. So our campaigns and background permissions syncers also live in `repo-updater`.
```

Reviewer: Any reason you choose to use "Code Host" not "code host" (which used elsewhere)?

Author: nope just oversight.

`gitserver` is a scaleable stateful service which clones git repositories and can run git commands against them. All data maintained on this service is from cloning an upstream repository. We shard the set of repositories across the gitserver replicas. The main RPC gitserver supports is `exec` which returns the output of the specified git command.

Note: The name `repo-updater` does not accurately capture what the service does. This is a historical artifact. We have not updated it due to the unneccessary operational burden it would put on our customers.

## Discovery

> Before we can clone a repository, we first must discover that is exists. This is configured by an administrator setting code host configuration. Typically a code host will have an API as well as git endpoints. A code host configuration typically will specify how to communicate with the API and which repositories to ask the API for. For example:

Reviewer: ```suggestion
Before we can clone a repository, we first must discover that is exists. This is configured by a site administrator setting code host configuration. Typically a code host will have an API as well as git endpoints. A code host configuration typically will specify how to communicate with the API and which repositories to ask the API for. For example:
```

``` json
{
  "url": "https://github.com",
  "token": "deadbeaf",
  "repositoryQuery": ["affiliated"],
}
```

This is a GitHub code host configuration for `github.com` using the private access token `deadbeaf`. It will ask GitHub for all affiliated repositories. Follow [`GithubSource.listRepositoryQuery`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/github.go#L612) to find the actual API call we do.

Discovering the repositories for each codehost/configuration is abstracted in the [`Sources interface`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/sources.go#L82:1).

``` go
// A Source yields repositories to be stored and analysed by Sourcegraph.
// Successive calls to its ListRepos method may yield different results.
type Source interface {
	// ListRepos sends all the repos a source yields over the passed in channel
	// as SourceResults
	ListRepos(context.Context, chan SourceResult)
	// ExternalServices returns the ExternalServices for the Source.
	ExternalServices() ExternalServices
}
```

## Syncing

> We keep a list of all repositories on Sourcegraph in the [`repos` table](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/frontend/db/schema.md#table-public-repo). This is so to provide a code host independent list of repositories on Sourcegraph that we can quickly query. `repo-updater` will periodically list all repositories from all sources and update the table. We need to list everything so we can detect which repositories to delete. See [`Syncer.Sync`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/syncer.go#L101) for details.

Reviewer: ```suggestion
We keep a list of all repositories on Sourcegraph in the [`repo` table](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/frontend/db/schema.md#table-public-repo). This is so to provide a code host independent list of repositories on Sourcegraph that we can quickly query. `repo-updater` will periodically list all repositories from all sources and update the table. We need to list everything so we can detect which repositories to delete. See [`Syncer.Sync`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/syncer.go#L101) for details.
```

The scheduler is divided into two parts:
- [`updateQueue`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/scheduler.go#L392:6) is a priority queue of repositories to clone/fetch on `gitserver`.
- [`schedule`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/scheduler.go#L567:6) which places repositories onto the `updateQueue` when it thinks it should be updated. This is what paces out updates for a repository. It contains heuristics such that recently updated repositories are more frequently checked.

Repositories can also placed onto the `updateQueue` if we receive a webhook indicating the repository has changed. (We don't by default setup webhooks when integrating into a code host). When a user directly visits a repository on Sourcegraph we also enqueue it for update.

> The [update scheduler](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/cmd/repo-updater/repos/scheduler.go#L165:27) has [`conf.GitMaxConcurrentClones`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v3.14.0/-/blob/schema/site.schema.json#L235-240) workers processing the `updateQueue` and issuing git clone/fetch commands.

Reviewer 2: Not sure if it is worth calling out here, but gitserver also enforces GitMaxConcurrentClones PER SHARD, because repo-updater is not the only codepath that calls into gitserver afaik. This means in aggregate it is possible (but perhaps not common) for Sourcegraph to be executing GITSERVER_REPLICA_COUNT * GitMaxConcurrentClones concurrent operations on a code host.

There is also one caveat for repo-updater. If a repo is being updated (not cloned) then that code path can return early (if the context deadline is exceeded) before the update finishes. To the extent this is true, even if repo-updater were the only codepath calling into gitserver, the number of concurrent operations could exceed GitMaxConcurrentClones (up to GitMaxConcurrentClones * GITSERVER_REPLICA_COUNT)

https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/gitserver/server/server.go#L416:18

Author: Good point, I will call this out. Also VPs aren't supposed to know this level of where the bodies are buried!
